output/shared/messagepacker.go

```diff
@@ -0,0 +1,142 @@
+package shared
+
+import (
+	"bytes"
+	"compress/gzip"
+
+	"github.com/relex/gotils/logger"
+	"github.com/relex/slog-agent/base"
+	"github.com/relex/slog-agent/defs"
+	"github.com/relex/slog-agent/util"
+)
+
+// gzipCompressionLevel for ForwardMessage.
+// BestSpeed uses 30% more space and roughly same percentage in time saving
+const gzipCompressionLevel = gzip.BestSpeed
+
+type PackerConfig struct {
+	MsgBufCapacity    int
+	ChunkMaxSizeBytes int
+	ChunkMaxRecords   int
+	ChunkIDSuffix     string
+	UseCompression    bool
+}
+
+// messagePacker writes incoming log messages into a batch (chunk) and encodes them to gzip
+type messagePacker struct {
+	logger            logger.Logger
+	useCompression    bool
+	reusedGzipBuffer  *bytes.Buffer // buffer for gzipWriter for log records
+	chunkIDGenerator  *chunkIDGenerator
+	currentChunk      *forwardMessageOpenChunk // current chunk before being made into final message
+	chunkMaxSizeBytes int
+	chunkMaxRecords   int
+	encoder           messageEncoder // output-specific chunk encoder
+}
+
+type forwardMessageOpenChunk struct {
+	ID             string       // chunk ID
+	GZIPWriter     *gzip.Writer // gzip writer can be nil if no compression, write to reusedGzipBuffer
+	NumRecords     int          // numbers of collected log records
+	NumStreamBytes int          // uncompressed length of written stream data
+}
+
+type messageEncoder interface {
```

Comment:
need to be `MessageEncoder` probably, since it's used from outside of the package